### ISSUE TITLE
Instability of Error Message Retrieval from Exceptions

### ISSUE THREAD

**Issue by ccamel** [ichiban/prolog#311]
## Issue

The process of retrieving the string representation of an Exception - `engine.Exception.Error()` - is found to be inconsistent across multiple calls. The first call returns the accurate error message, but the second and subsequent calls return the string `...`.

The following minimal code demonstrates the issue:

```go
package main

import (
    "fmt"
    "github.com/ichiban/prolog"
    "github.com/ichiban/prolog/engine"
)

func Foo(_ *engine.VM, foo engine.Term, cont engine.Cont, env *engine.Env) *engine.Promise {
    return engine.Error(engine.TypeError(engine.NewAtom("foo"), foo, env))
}

func main() {
    i := &prolog.Interpreter{}
    i.Register1(engine.NewAtom("foo"), Foo)

    sol := i.QuerySolution("foo(What).")
    err := sol.Err()
    if err != nil {
        fmt.Printf("Error: %s\n", err.Error()) // 👉 Error: error(type_error(foo,_4),foo/1)
        fmt.Printf("Error: %s\n", err.Error()) // 👉 Error: ...
    }
}
```

## Analysis

The issue appears to originate from the global variable [engine.defaultWriteOptions](https://github.com/ichiban/prolog/blob/v1.1.3/engine/term.go#L60) used for managing the [compounds already visited](https://github.com/ichiban/prolog/blob/v1.1.3/engine/compound.go#L57) in the string conversion process of the `engine.Exception`.

In the first call, the table of visited compounds is populated, leading to the accurate error message. However, in subsequent calls, as the compound has already been visited, the writer generates the string `...` instead. 😞

## Impact

This side effect is particularly unexpected as it's common to get consistent string returns from exceptions.

Additionally, the effect is quite subtle in the context of our unit tests using [smartystreets/goconvey](https://github.com/smartystreets/goconvey), where assertions are structured as follows:

```go
  So(err, ShouldNotBeNil)
  So(err.Error(), ShouldEqual, wanted.err.Error())
```

In this scenario, the first call (which is introspective) generates an accurate error `string`. However, on the second call, the output becomes `...`, which is unfortunately incorrect. 😥

**Comment by ichiban** [comment]
Could you try v1.1.1 or newer, please? I think it's fixed in https://github.com/ichiban/prolog/pull/298.
Let me know if the problem persists!

**Comment by ccamel** [comment]
@ichiban, yes it's fixed in version v1.1.3. :relieved: 

Thanks!
